### ISSUE TITLE
Optimizers not working in callbacks when multiple optimizers in training

### ISSUE THREAD

**Issue by YannDubs** [PyTorchLightning/pytorch-lightning#4955]
## 🐛 Bug

If the trainer uses multiple optimizers (e.g. GANs) then you cannot use optimizers in callbacks (e.g. `pl_bolts.callbacks.ssl_online.SSLOnlineEvaluator`) because when using multiple optimizers `toggle_optimizer` is called during training, but `requires_grad` are not set to True at the end of training (so gradients for callback optimizers are not computed).

## Please reproduce using the BoringModel and post here

https://colab.research.google.com/drive/1ryy-FwUhQmTvgBvjF9cC3zVmrGFIAcxr?usp=sharing

### Expected behavior

At the end of training (but before callbacks) all `requires_grad` should be set to true again, so that users can have optimizers in callbacks. If not `pl_bolts.callbacks.ssl_online.SSLOnlineEvaluator` does not work when using multiple optimizers.

### Environment

pytorch-lightning master (same bug with pip install)

**Comment by tchaton** [comment]
Hey @YannDubs,

Could you provide reproduce this behaviour using https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pl_examples/bug_report_model.py.
I am not sure to fully understand what you are trying to do.

Best regards,
Thomas Chaton

**Comment by edenlightning** [comment]
Feel free to reopen with reproducible code!
